--- installOpenFOAM-master/install.sh	2020-07-28 13:25:56.000000000 +0900
+++ cc-openfoam/v1906/installOpenFOAM/install.sh	2021-04-23 12:18:43.464987800 +0900
@@ -1,7 +1,7 @@
 #!/bin/bash
 
 [ $BASH_SOURCE ] && INSTALLOPENFOAM_DIR=$(\cd ${BASH_SOURCE%/*} && \pwd -P) || \
-INSTALLOPENFOAM_DIR=$HOME/OpenFOAM/installOpenFOAM
+INSTALLOPENFOAM_DIR=/mnt/share/OpenFOAM/v1906/installOpenFOAM
 
 #
 # Usage
@@ -84,6 +84,9 @@ download_and_extract_source()
 		fi
 	    done
 	    (
+		echo "\$FOAM_INSTALL_DIRECTORY/\$PACKAGE : $FOAM_INSTALL_DIRECTORY/$PACKAGE"
+		echo "\$FOAM_INSTALL_DIRECTORY : $FOAM_INSTALL_DIRECTORY"
+		echo "\$PACKAGE : $PACKAGE"
 		cd $FOAM_INSTALL_DIRECTORY/$PACKAGE
 		git init
 		git config user.name "The Open CAE Society of Japan"
@@ -166,13 +169,21 @@ $MPC_PACKAGE"
     esac
 
     case "$FOAM_VERSION" in
-	v1712*|v1806*|v1812*|v1906*|v2006*)
+	v17*|v1806*|v1812*|v1906*|v1912*|v2006*)
 	    foamConfigurePathsOptions="\
 $foamConfigurePathsOptions \
 -kahip $KAHIP_PACKAGE"
 	    ;;
     esac
 
+    case "$FOAM_VERSION" in
+	v1706*)
+	    foamConfigurePathsOptions="\
+$foamConfigurePathsOptions \
+$KAHIP_PACKAGE"
+	    ;;
+    esac
+
     (cd $WM_PROJECT_DIR
 	bin/tools/foamConfigurePaths $foamConfigurePathsOptions
     )
@@ -725,6 +736,7 @@ build_ThirdParty()
 	source $WM_PROJECT_DIR/etc/bashrc $foam_settings
 	chmod +x make[a-zA-Z0-9]*
 	[ -n "$PLUS_VERSION" ] && options="-k"
+        sed -i".org" -e s/'--enable-mpi-fortran=none'/'--enable-mpi-fortran=none --without-ucx'/g makeOPENMPI
 	./Allwmake $options
     )
 }
